apps/cowswap-frontend/src/modules/appData/utils/buildAppData.ts

```diff
@@ -125,12 +138,61 @@ function mergeHooks(
   if (!hooks1 && hooks2) return hooks2
 
   if (hooks1 && hooks2) {
+    const pre = (hooks1.pre || []).concat(hooks2.pre || [])
+    const post = (hooks1.post || []).concat(hooks2.post || [])
+
+    if (!pre && !post) {
+      // Avoid empty hooks
+      console.log(`bug:mergeHooks empty pre and post`, hooks1, hooks2)
+      return undefined
+    }
+
+    console.log(`bug:mergeHooks has either pre, post or both`, hooks1, hooks2, pre, post)
+
     return {
       version: hooks1.version,
-      pre: (hooks1.pre || []).concat(hooks2.pre || []),
-      post: (hooks1.post || []).concat(hooks2.post || []),
+      // Remove the ones that are empty here too
+      ...(pre ? { pre } : undefined),
+      ...(post ? { post } : undefined),
     }
   }
 
   return undefined
 }
+
+function removeDuplicatedHook<T extends PreHooks | PostHooks>(hooks: T | undefined): T | undefined {
+  if (!hooks || hooks.length < 2) {
+    console.log(`bug:removeDuplicatedHook no hooks or only 1`, hooks?.length)
+    return hooks
+  }
+
+  const duplicatedIndices: Set<number> = new Set()
+
+  // Check all hooks against each other to identify duplicates
+  for (let i = 0; i < hooks.length; i++) {
+    if (duplicatedIndices.has(i)) {
+      console.log(`bug:removeDuplicatedHook skipping duplicate`, i)
+      continue
+    }
+
+    const hookA = hooks[i]
+
+    for (let j = i + 1; j < hooks.length; j++) {
+      const hookB = hooks[j]
+
+      if (hooksAreEqual(hookA, hookB)) {
+        console.log(`bug:removeDuplicatedHook found duplicate`, hookB)
+        duplicatedIndices.add(j)
+      }
+    }
+  }
+
+  return duplicatedIndices.size ? (hooks.filter((hook, index) => !duplicatedIndices.has(index)) as T) : hooks
```

Comment:
Is this too complicated?
I think this could be simplified, will try to refactor it.

Reply:
More that complicated... what I don't understand:
- Isn't the issue that we add a second permit when we have one? 
- This logic is comparing for equality, but wouldn't it be also wrong to have 2 permits for the same token? (maybe we should classify the hooks, and really enforce this rules) --> no more than one permit hook (no matter if the call-data is the same or not

Reply:
The problem is that hooks don't have any identification in them.
All we have is: callData, target and gas.

A more convoluted solution would be to add such info and remove them before the order is quoted/signed as the extra data would lead to different hashes. And/or change the appData to include the info there.
I'm not very found of either solution to be honest.

Reply:
That is my suggestion, to model it in our app. Not into the CowHook itself, but into the CowHookDetails

something like:
```ts
export interface CowHookAppDetailsBase {
  uuid: string
  hook: CowHook
  outputTokens?: CurrencyAmount<Currency>[]
}

export interface CowHookAppDetails extends CowHookAppDetailsBase {
  type: 'hook-stoe'  
  dapp: HookDapp  
}

export interface CowHookInternalDetails extends CowHookAppDetailsBase {
  type: 'permit' // || other internal hooks
}
```

This way, we just add a CowHoowDetails into the state when we add the permit (instead of just the CowHook data) so we have the full context.

This will be relevant too if we do the bridging as it will likely be a hook we cook ourselves too